### PR TITLE
usability improvements with how base options are registered

### PR DESCRIPTION
**Pull request by busterb**
This adds named parameters for all of the current array-index based options. It also allows specifying the description as the 2nd parameter, allowing the 'required' parameter to be implicitly false (the most common value).

A simple parameter like:

```
     OptAddress.new('ReverseListenerBindAddress',
       [false, 'The specific IP address to bind to on the local system']),
```

Can now be rewritten as:

```
     OptAddress.new('ReverseListenerBindAddress',
       'The specific IP address to bind to on the local system'),
```

More complex options are also now easier to read:

```
     OptString.new(
       'HttpUserAgent',
       'The user-agent that the payload should use',
       default: Rex::UserAgent.shortest,
       aliases: ['MeterpreterUserAgent']
     ),
```

This also makes dealing with enums easier because default is implicitly the first element, unless specified. This:

```
      OptEnum.new('PayloadProxyType',
        [true, 'The proxy type, HTTP or SOCKS', 'HTTP', ['HTTP', 'SOCKS']]),
```

    becomes:

```
      OptEnum.new('HttpProxyType',
        'The proxy type, HTTP or SOCKS', required: true, enums: ['HTTP', 'SOCKS'])
```

This maintains full backward compatibility with existing code, so you don't have to use this (and to convert everything would be a rather unrealistic task!)

Note there are some whitespace changes here, which you can filter out with https://github.com/rapid7/metasploit-framework/pull/9000/files?w=1

### CODE REVIEW

File: lib/msf/core/opt.rb [hunk 1]
```diff
@@ -53,20 +52,10 @@ def self.RPORT(default=nil, required=true, desc="The target port")
...
-      alias builtin_chost CHOST
-      alias builtin_cport CPORT
-      alias builtin_lhost LHOST
-      alias builtin_lport LPORT
-      alias builtin_proxies Proxies
-      alias builtin_rhost RHOST
-      alias builtin_rport RPORT
+      Msf::OptEnum.new('SSLVersion',
+        'Specify the version of SSL/TLS to be used (Auto, TLS and SSL23 are auto-negotiate)',
+        enums: ['Auto', 'SSL2', 'SSL3', 'SSL23', 'TLS', 'TLS1', 'TLS1.1', 'TLS1.2']
```
Comment (sempervictus) [inline lib/msf/core/opt.rb]: The enum should probably be dynamic based on what openssl can support - pretty sure we can't sslv2.

File: lib/msf/core/opt_base.rb [hunk 2]
```diff
@@ -22,15 +22,39 @@ class OptBase
...
       self.advanced = false
       self.evasion  = false
-      self.required = attrs[0] || false
-      self.desc     = attrs[1]
-      self.default  = attrs[2]
-      self.enums    = [ *(attrs[3]) ].map { |x| x.to_s }
-      regex_temp    = attrs[4] || nil
+      self.aliases  = aliases
+
+      if attrs.is_a?(String) || attrs.length == 0
```
Comment (sempervictus) [inline lib/msf/core/opt_base.rb]: Nitpick: Brixen once told me that :is_a? is costly as it hits up the GC or some other low level VM component. AFAIK, its more efficient to call :responds_to?(:upcase)

File: lib/msf/core/opt_base.rb [hunk 3]
```diff
@@ -45,35 +69,34 @@ def initialize(in_name, attrs = [], aliases: [])
...
         end
       end
-      self.aliases = aliases
     end
 
     #
     # Returns true if this is a required option.
     #
     def required?
-      return required
```
Comment (sempervictus) [inline lib/msf/core/opt_base.rb]: Do we have an official policy on this? I'm a fan of explicit return declarations when there's any question about what we may kick back, but also for visibility. I can't argue with this case, if the reader doesn't see the returns here they probably need to brush up on their Ruby, but we do have people coming on from other languages...

File: lib/msf/core/opt_bool.rb [hunk 4]
```diff
@@ -1,48 +1,33 @@
...
+    def valid?(value, check_empty: true)
+      return false if check_empty && empty_required_value?(value)
+      return true if value.nil? && !required?
 
-  def normalize(value)
-    if(value.nil? or value.to_s.match(TrueRegex).nil?)
-      false
-    else
-      true
+      !(value.nil? ||
```
Comment (sempervictus) [inline lib/msf/core/opt_bool.rb]: ActiveSupport gives us :blank? Which, if memory serves you taught me (or maybe wvu?).

File: lib/msf/core/opt_enum.rb [hunk 5]
```diff
@@ -1,48 +1,45 @@
...
-  def desc
-    if self.enums
-      str = self.enums.join(', ')
+      !value.nil? && enums.include?(value.to_s)
     end
-    "#{self.desc_string || ''} (Accepted: #{str})"
-  end
 
+    def normalize(value = self.value)
+      !valid?(value) ? nil : value.to_s
```
Comment (sempervictus) [inline lib/msf/core/opt_enum.rb]: "You're my ternary hero" - show that to Egypt, watch him turn purple.

**Review by sempervictus — COMMENTED** [review]
This needs more eyes, sorry - I've a serious cold and am half brain dead (hence the post time), and it took a bit of effort to read through this. I'll merge in locally and do a merge review too, but it seems viable from first pass. None of the inline comments are blockers.
The direction is very welcome, over the years I've forced myself to get used to the existing approach but have always thought it to be a bit clunky. Thanks for tackling this boss.

File: lib/msf/core/opt_bool.rb (hunk 4 above)
Comment (wvu-r7) [reply lib/msf/core/opt_bool.rb]: Yup, `blank?` will tell you if it's `nil`, too.

File: lib/msf/core/opt_enum.rb (hunk 5 above)
Comment (wvu-r7) [reply lib/msf/core/opt_enum.rb]: Funny enough, the style guide now endorses ternary. ;)

Comment (sempervictus) [reply lib/msf/core/opt_enum.rb]: To quote the bug from family guy who started fires - "good, good."

Comment (bcook-r7) [reply lib/msf/core/opt_enum.rb]: life is too short, let's make cool stuff

File: lib/msf/core/opt.rb (hunk 1 above)
Comment (bcook-r7) [reply lib/msf/core/opt.rb]: Sure, sounds neat.

File: lib/msf/core/opt_base.rb (hunk 3 above)
Comment (bcook-r7) [reply lib/msf/core/opt_base.rb]: Was just trying to be consistent. We had a mix before.

File: lib/msf/core/opt_enum.rb (hunk 5 above)
Comment (sempervictus) [reply lib/msf/core/opt_enum.rb]: Hey I'm all for this, its readable and neat. If the method name was not explicit, this might suck, but as is it reads out loud in plain English.

**Comment by busterb** [comment]
Building our own OpenSSL gem makes me both intrigued and nauseated. I poked lightly at the one that rubinius packages, but if someone can make one (and figure out how to push prebuilt binaries), I'm game.

File: lib/msf/core/opt_base.rb (hunk 2 above)
Comment (acammack-r7) [reply lib/msf/core/opt_base.rb]: Not seeing this in the synthetic case on ruby 2.3:
```
$ time ruby -e '10_000_000.times { "arg".is_a? String }

real    0m0.777s
user    0m0.774s
sys     0m0.004s

$ time ruby -e '10_000_000.times { "arg".respond_to? :upcase }

real    0m1.019s
user    0m1.007s
sys     0m0.014s
```

**Comment by sempervictus** [comment]
Brixen did rubysl to reduce the codebase inside rbx, but as of mri 2.4 I believe there is now a gem for all. Will dig in.

-------- Original Message --------

File: lib/msf/core/opt_enum.rb (hunk 5 above)
Comment (acammack-r7) [reply lib/msf/core/opt_enum.rb]: Could we at least get rid of the unnecessary `!` and leave it as `valid?(value) ? value.to_s : nil`?

Comment (egypt) [reply lib/msf/core/opt_enum.rb]: /me turns purple

**Comment by busterb** [comment]
rebased and addressed all of the addressable comments here. Sadly, Ruby returns a list of SSL methods which have no basis in reality, so you have to manually check them all anyway.

**Comment by busterb** [comment]
I also tried using activesupport, but we don't require it at the right level for it to be usable in this leaf code. Didn't seem worth it to require over something so minor.

File: lib/msf/core/opt_bool.rb (hunk 4 above)
Comment (busterb) [reply lib/msf/core/opt_bool.rb]: sorry, it doesn't actually work here (tried and failed)

**Comment by Barthyr** [comment]
w

**Comment by tdoan-r7** [comment]
# Release Notes
Named parameters have been added for all of the current array-index based options.
